src/SpriteDecoders/Ambrosia-btSP-HrSp-SprT.cc

```diff
@@ -1,202 +1,321 @@
+    size_t header_size) {
+  if (header_size < 8) {
+    throw logic_error("header size is too small");
+  }
+  if (header_size & 3) {
+    throw logic_error("header size must be a multiple of 4");
+  }
+  if (data.size() < header_size + 4) {
+    throw invalid_argument("not enough data");
+  }
+  if (data.size() & 3) {
+    throw invalid_argument("size must be a multiple of 4");
+  }
+
+  StringReader r(data.data(), data.size());
+  r.go(4);
+  uint16_t height = r.get_u16b();
+  uint16_t width = r.get_u16b();
+  r.go(header_size);
+
+  // Known commands:
+  // 00 00 00 00 - end
+  // 01 XX XX XX - row frame (the next row begins when we've executed this many more bytes)
+  // 02 XX XX XX - write X bytes to current position
+  // 03 XX XX XX - write X transparent bytes
+
+  Image ret(width, height, true);
+  size_t x = 0, y = 0;
+  size_t next_row_begin_offset = static_cast<size_t>(-1);
+  while (!r.eof()) {
+    if (r.where() == next_row_begin_offset) {
+      x = 0;
+      y++;
+    }
+
+    uint8_t cmd = r.get_u8();
+    switch (cmd) {
+
+      case 0:
+        if (r.get_u24b() != 0) {
+          throw runtime_error("end-of-stream command with nonzero argument");
+        }
+        if (!r.eof()) {
+          throw runtime_error("end-of-stream command not at end of stream");
+        }
+        break;
+
+      case 1:
+        next_row_begin_offset = r.get_u24b();
+        next_row_begin_offset += r.where();
+        break;
+
+      case 2: {
+        uint32_t count = r.get_u24b();
+        for (uint32_t z = 0; z < count; z++) {
+          uint8_t v = r.get_u8();
+          auto c = clut.at(v).c.as8();
+          ret.write_pixel(x, y, c.r, c.g, c.b);
+          x++;
+        }
+        // Commands are padded to 4-byte boundary
+        while (count & 3) {
+          r.get_u8();
+          count++;
+        }
+        break;
+      }
+
+      case 3: {
+        uint32_t count = r.get_u24b();
+        for (uint32_t z = 0; z < count; z++) {
+          ret.write_pixel(x, y, 0x00, 0x00, 0x00, 0x00);
+          x++;
+        }
+        break;
+      }
+
+      default:
+        throw runtime_error(string_printf("unknown command: %02hhX", cmd));
+    }
+  }
+
+  return ret;
+}
+
+vector<Image> decode_SprD(const string &data, const vector<ColorTableEntry> &clut,
+                          size_t header_size)
+{
+  if (header_size < 8)
+  {
+    throw logic_error("header size is too small");
+  }
+  if (header_size & 3)
+  {
+    throw logic_error("header size must be a multiple of 4");
+  }
+  if (data.size() < header_size + 4)
+  {
+    throw invalid_argument("not enough data");
+  }
+  if (data.size() & 3)
+  {
+    throw invalid_argument("size must be a multiple of 4");
+  }
+
+  vector<Image> imageVector;
+
+  // Initializing starting image
+  StringReader r(data.data(), data.size());
+  r.go(4);
+  uint16_t height = r.get_u16b();
+  uint16_t width = r.get_u16b();
+  r.go(header_size);
```

Comment:
The header_size argument in decode_HrSp only exists because there are two slightly different formats out there (Harry and Bonkheads use the same commands, but the start offset is different). Since Slithereens is the only game known to use this format, it would be fine to just hardcode 20 here instead.